### PR TITLE
Generalized indentation linter

### PR DESCRIPTION
**Pull request by flying-sheep**
(reopened #139 from a new branch)

Fixes part of #135

the idea is to grab as many units of valid indentation from the start of the line (1 tab, or n spaces for a fixed n), and look if there’s still spaces or tabs after that. if so, return a lint.

### CODE REVIEW

**Comment by flying-sheep** [comment]
yuss, it works!

File: R/indentation_linters.R [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#' @describeIn linters check that a certain kind of indentation is used
+#' @export
+indentation_linter <- function(spaces, linter_name = "indentation_linter") function(source_file) {
+  if (spaces %==% FALSE || spaces %==% 0L) {
+    allowed <- "\t"
+    message <- "Use tabs to indent."
+  } else {
+    nspc <- spaces  # else rex' internal 'spaces' is used!
+    allowed <- rex(n_times(" ", nspc))
+    message <- sprintf("Use %s spaces to indent.", digit_to_word(spaces))
```
Comment (jimhester) [inline R/indentation_linters.R]: Rather than writing the word out I think we should just use the number directly.

Comment (flying-sheep) [inline R/indentation_linters.R]: i wanted to preserve the nice error message, but OK!

**Comment by dgkf** [comment]
Apologies for being ~4 years late to this, but I recently took the plunge into linter writing and wrote a similar general indentation linter. I started off just writing the linters as a fun exploratory project to learn more about the internals of the lintr package, so forgive me for not doing my due diligence to see if I was reinventing any wheels - my priority was just learning how the package works. 

In my version I took a bit of a different approach - instead of using regex, I used the `parsed_content` to get the indentation level per line and throw Lints when a nested expression isn't indented relative to the parent expression (as well as when closing curly braces don't return to the parent indentation).

Code available in my fork [dgkf/lintr@argument_wrapping_lintr](https://github.com/dgkf/lintr/blob/dgkf/argument_wrapping_linter/R/indentation_linter.R). There are a couple others I've played around with - one to lint arguments on newlines and another to match the opening and closing of brackets on the same lines - just some of my stylistic preferences that I wanted to toy around with to build some chops with this package.

There's still some cruft to the code that I need to clean up, but wanted to share my effort in case it's useful. If it looks remotely helpful, let me know and I'd be happy to help get it in a shape that would be more actionable.

**Comment by jimhester** [comment]
@dgkf Your approach sounds promising to me! If you would like to open a new PR we can refine it there.

**Comment by flying-sheep** [comment]
Yeah, would be great if you could take this over, I don’t work on R much these days.

**Comment by MichaelChirico** [comment]
Closing this. @dgkf please feel free to file your work as a new PR.

**Pull request closed without merging**